### PR TITLE
Pointcloud improvement

### PR DESCRIPTION
**Pull request by peppsac**
2 commits improving pointcloud support for itowns.

The first one fixes the point budget handling for .bin format.

The second reworks SSE: it's similar to #674 but simpler and only for pointclouds.

Full commits message:
   - refactor(pointcloud): rework point budget implementation
    
    Since .bin and .cin have different repartitions of the points inside a node,
    they must use different algorithms.
    For .cin we want to draw a %-age of each node because points are evenly
    distributed.
    For .bin, we sort the nodes by 'importance', and stop drawing nodes when we
    have enoug points.
   - refactor(pointcloud): simplify SSE computation for pointcloud
    
    Project average point spacing on screen and compare to sse threshold.
    It's a simplification of the approach proposed in #674.
    
    This allows to drop a dependency (convexhull) and brings the SSE
    computation for pointcloud closer to what is done for other geometries.

Edit:
The SSE value computed here is the estimation of the maximum distance between 2 points in a given node.
PotreeConverter guarantees that the distance between each pair of points is >= `metadata.spacing / node.level` in all nodes. This is easy to visualize if you display successive children (using the 'sticky' debug tool):
distances

The SSE is the estimation of this distance projected on the screen.

### CODE REVIEW

**Comment by peppsac** [comment]
Edited description + added image

File: src/Process/PointCloudProcessing.js
Comment (mbredif) [inline src/Process/PointCloudProcessing.js]: Maybe you could add a comment here stating what -1 means

File: src/Process/PointCloudProcessing.js
Comment (mbredif) [inline src/Process/PointCloudProcessing.js]: You could link here to an existing technical paper to explain the derivation

File: src/Process/PointCloudProcessing.js
Comment (mbredif) [inline src/Process/PointCloudProcessing.js]: Why do you divide by distance ?

File: src/Process/PointCloudProcessing.js
Comment (mbredif) [inline src/Process/PointCloudProcessing.js]: I would rather have the processing be agnostic of formats, nut depend rather on format properties or strategies. Could not you have reordered bin files as in cin files, or the other way around?

**Review by mbredif — COMMENTED** [review]
Looks good and simplifies things, thanks !
I added a few comments to improve it even further :)
(I have not tested it though)

File: src/Process/PointCloudProcessing.js
Comment (mbredif) [reply src/Process/PointCloudProcessing.js]: What about renaming customBinFormat to something like prefixSubsampling ?

File: src/Process/PointCloudProcessing.js
Comment (peppsac) [reply src/Process/PointCloudProcessing.js]: To load closest node first.

File: src/Process/PointCloudProcessing.js
Comment (peppsac) [reply src/Process/PointCloudProcessing.js]: I added a `layer.supportsProgressiveDisplay` property. (Note that the rest of the file relies on the potreeconverter format, there's more work to do to be able to use this processing code on any pointcloud source)

File: src/Process/PointCloudProcessing.js
Comment (peppsac) [reply src/Process/PointCloudProcessing.js]: Done

File: src/Process/PointCloudProcessing.js
Comment (peppsac) [reply src/Process/PointCloudProcessing.js]: Done

File: src/Process/PointCloudProcessing.js
Comment (mbredif) [reply src/Process/PointCloudProcessing.js]: If I understand correctly, minimizing SSE is what you want eventually (as things settle), but you prefer to bias the ordering of the fetches towards closer nodes, even if those have a smaller SSE. Is that correct ? I think a comment should be placed here for future reference (eg, we might want to enable the customization of this strategy)

**Comment by gchoqueux** [comment]
this commit : **refactor(pointcloud): rework point budget implementation**
Does it allow  `PointCloud` to be fully compliant with the data potree (from [potree converter](https://github.com/potree/PotreeConverter))? Especially in the organization of points in the data structure

File: src/Process/PointCloudProcessing.js
Comment (mbredif) [inline src/Process/PointCloudProcessing.js]: is it equivalent to returning Infinity ? (which would be parameterless and thus more explicit)

**Comment by peppsac** [comment]
> this commit : refactor(pointcloud): rework point budget implementation
Does it allow PointCloud to be fully compliant with the data potree (from potree converter)? Especially in the organization of points in the data structure

Yes.

**Review by mbredif — COMMENTED** [review]
LGTM

**Comment by gchoqueux** [comment]
Ok thanks, it's good to be completely compliant with Potree data

File: src/Process/PointCloudProcessing.js
Comment (peppsac) [reply src/Process/PointCloudProcessing.js]: Thanks for the review. I addressed both comments.

**Review by mbredif — APPROVED** [review]
LGTM

**Comment by peppsac** [comment]
Thanks for the review @mbredif - and welcome to the @iTowns/reviewers team !

**Comment by autra** [comment]
> Thanks for the review @mbredif - and welcome to the @iTowns/reviewers team !

actually he just has write access. The convention is : a review from someone from itowns-reviewers group. 

Nevermind, I'll make some comments on this if needed and we'll do a followup, ok for you?

**Comment by peppsac** [comment]
Oops my bad. I'm reverting this PR then, until your approval.

**Comment by peppsac** [comment]
PR reverted in #745 and reopened as #746.

**Pull request merged**
